Mathlib/Order/PrimeSeparator.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2024 Sam van Gool. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sam van Gool
+-/
+
+import Mathlib.Order.PrimeIdeal
+import Mathlib.Order.Zorn
+
+
+/-!
+# Separating prime filters and ideals
+
+In a bounded distributive lattice, if $F$ is a filter, $I$ is an ideal, and $F$ and $I$ are
+disjoint, then there exists a prime ideal $J$ containing $I$ with $J$ still disjoint from $F$.
+This theorem is a crucial ingredient to [Stone's][Sto1938] duality for bounded distributive
+lattices. The construction of the separator relies on Zorn's lemma.
+
+## Tags
+
+ideal, filter, prime, distributive lattice
+
+## References
+
+* [M. H. Stone, Topological representations of distributive lattices and Brouwerian logics
+(1938)][Sto1938]
+-/
+
+
+universe u
+variable {α : Type*}
+
+open Order Ideal Set
+
+variable [DistribLattice α] [BoundedOrder α]
+
+variable {F : PFilter α} {I : Ideal α}
+
+namespace DistribLattice
+
+
+lemma mem_ideal_sup_principal (a b : α) (J : Ideal α) : b ∈ J ⊔ principal a ↔ ∃ j ∈ J, b ≤ j ⊔ a :=
+  ⟨fun ⟨j, ⟨jJ, _, ha', bja'⟩⟩ => ⟨j, jJ, le_trans bja' (sup_le_sup_left ha' j)⟩,
+    fun ⟨j, hj, hbja⟩ => ⟨j, hj, a, le_refl a, hbja⟩⟩
+
+theorem prime_ideal_of_disjoint_filter_ideal (hFI : Disjoint (F : Set α) (I : Set α)) :
+    ∃ J : Ideal α, (IsPrime J) ∧ I ≤ J ∧ Disjoint (F : Set α) J := by
+
+  -- Let S be the set of proper ideals containing I and disjoint from F
+  set S : Set (Set α) := { J : Set α | IsIdeal J ∧ I ≤ J ∧ Disjoint (F : Set α) J }
+
+  -- Then I is in S...
+  have IinS : ↑I ∈ S := by
+    refine ⟨Order.Ideal.isIdeal I, by trivial⟩
+
+  -- ...and S contains upper bounds for any non-empty chains.
+  have chainub : ∀ c ⊆ S, IsChain (· ⊆ ·) c → c.Nonempty →  ∃ ub ∈ S, ∀ s ∈ c, s ⊆ ub := by
+    intros c hcS hcC hcNe
+    use sUnion c
+    refine ⟨?_, fun s hs ↦ le_sSup hs⟩
+    simp only [le_eq_subset, mem_setOf_eq, disjoint_sUnion_right, S]
+    let ⟨J, hJ⟩ := hcNe
+    refine ⟨Order.isIdeal_sUnion_of_isChain (fun _ hJ ↦ (hcS hJ).1) hcC hcNe,
+            ⟨le_trans (hcS hJ).2.1 (le_sSup hJ), fun J hJ ↦ (hcS hJ).2.2⟩⟩
+
+  -- Thus, by Zorn's lemma, we can pick a maximal ideal J in S.
+  have zorn := zorn_subset_nonempty S chainub I IinS
+  have hJ := Exists.choose_spec zorn
+  set Jset := Exists.choose zorn
+  let ⟨⟨Jidl, IJ, JF⟩, ⟨_, Jmax⟩⟩ := hJ
```

Comment:
Oh wait, I'm a silly goose. `rcases`/`obtain` can do this directly since we're making a `Prop`!
```suggestion
  obtain ⟨Jset, ⟨Jidl, IJ, JF⟩, ⟨_, Jmax⟩⟩ := zorn_subset_nonempty S chainub I IinS
```

Reply:
Great.